proposals/NNNN-ComparisonReform.md

```diff
@@ -0,0 +1,531 @@
+with Equatable's Substitutability requirement. `-0` and `+0` have different
+behaviours: `1/-0 = -Inf` while `1/+0 = +Inf`. The main problem this can lead
+to is that a keyed collection may have two "0" entries. In practice this
+probably won't be a problem because it's fairly difficult for the same
+algorithm to produce both `-0` and `+0`. Any algorithm that does is also
+probably concerned with the fact that `1.0E-128` and `2.0E-128` are considered
+distinct values.
+
+Note: IEEE specifies several other potential total orderings: level 3, level 4,
+and the totalOrder predicate. For our purposes, these orderings are too
+aggressive in distinguishing values that are semantically equivalent in Swift.
+For most cases, the relevant issue is that they distinguish different encodings
+of NaN. For more exotic encodings that don't guarantee normalization, these
+predicates also consider `10.0e0 < 1.0e1` to be true. An example where this can
+occur is *IEEE-754 decimal coded floating point*, which FloatingPoint is
+intended to support.
+
+We will then make the comparison operators (`<`, `<=`, `==`, `!=`, `>=`, `>`)
+dispatch to one of `compare(_:)` or FloatingPoint's IEEE comparison methods
+(`isLess`, `isEqual`, `isLessThanOrEqualTo`) based on the context.
+
+* If the context knows the type is FloatingPoint, then level 1 ordering will be
+  used.
+* If the context only knows the type is Comparable or Equatable, then level 2
+  ordering will be used.
+
+This results in code that is explicitly designed to work with FloatingPoint
+types getting the expected IEEE behaviour, while code that is only designed to
+work with Comparable types (e.g. `sort` and `Dictionary`) gets more reasonable
+total ordering behaviour.
+
+To clarify: `Dictionary` and `sort` won't somehow detect that they're being
+used with `FloatingPoint` types and use level 1 comparisons. Instead they will
+unconditional use level 2 behaviour. For example:
+
+```swift
+let nan = 0.0/0.0
+
+func printEqual<T: Equatable>(_ x: T, _ y: T) {
+  print(x == y)
+}
+
+func printEqualFloats<T: FloatingPoint>(_ x: T, _ y: T) {
+  print(x == y)
+}
+
+print(nan == nan)          // false, (concrete)
+printEqual(nan, nan)       // true,  (generic Equatable but not FloatingPoint)
+printEqualFloats(nan, nan) // false, (generic FloatingPoint)
+```
+
+If one wishes to have a method that works with all Equatable/Comparable types,
+but uses level 1 semantics for FloatingPoint types, then they can simply
+provide two identical implementations that differ only in the bounds:
+
+```swift
+let nan = 0.0/0.0
+
+func printEqual<T: Equatable>(_ x: T, _ y: T) {
+  print(x == y)
+}
+
+func printEqual<T: FloatingPoint>(_ x: T, _ y: T) {
+  print(x == y)
+}
+
+printEqual(0, 0)           // true (integers use `<T: Equatable>` overload)
+printEqual(nan, nan)       // false (floats use `<T: FloatingPoint>` overload)
+```
+
+As a result of this change, hashing of floats must be updated to make all NaNs
```

Comment:
This seems potentially problematic.

I have a radical question: Should `FloatingPoint` types even be `Hashable` and `Equatable`? It seems to me that using `hashValue` or `==` on a `Float` is almost always the wrong thing to do.

Radical design, if only just to be refuted:
```swift
protocol Orderable {
  func <(lhs: Self, rhs: Self) -> Bool
}

protocol Equatable {
  func ==(lhs: Self, rhs: Self) -> Bool
}

protocol Comparable: Orderable, Equatable {
  // Requires total ordering
  func compare(lhs: Self, rhs: Self) -> ComparisonResult
}
extension Comparable {
  func ==(lhs: Self, rhs: Self) -> Bool {
    return compare(lhs, rhs) == .same
  }
    
  func <(lhs: Self, rhs: Self) -> Bool {
    return compare(lhs, rhs) == .ascending
  }
}
```
Here, it is not necessary that an `Orderable` type (like `Float`) have total ordering semantics. Instead, `Comparable` adds this requirement as well as `Equatable` inheritance. `Float` could provide functions like `Float.compare(a, b, semantics: .level2)` for when comparisons are necessary, but they would not be part of the protocol.

I realize this is all probably too source-breaking to even consider, but I thought I'd ask since I'm wondering what your thoughts are on this.